File: app/src/main/java/com/mewo/reader/data/Chapters.kt
package com.mewo.reader.data

fun List<FeedPost>.headingPosts(): List<FeedPost> =
    filter { it.kind == PostKind.Heading }

/**
 * Index in [headings] of the last heading at or before [visibleIndex] in [posts].
 * Used so the chapter sheet can mark and scroll to where you already are.
 */
fun currentHeadingIndex(
    posts: List<FeedPost>,
    headings: List<FeedPost>,
    visibleIndex: Int,
): Int {
    if (headings.isEmpty()) return 0
    val indexById = HashMap<String, Int>(posts.size)
    posts.forEachIndexed { index, post -> indexById[post.id] = index }
    val limit = visibleIndex
    return headings.indices.lastOrNull { i ->
        (indexById[headings[i].id] ?: Int.MAX_VALUE) <= limit
    } ?: 0
}

File: ios/Mewo/Data/Chapters.swift
import Foundation

extension Array where Element == FeedPost {
    func headingPosts() -> [FeedPost] {
        filter { $0.kind == .heading }
    }
}

/// Index in headings of the last heading at or before visibleIndex in posts.
func currentHeadingIndex(posts: [FeedPost], headings: [FeedPost], visibleIndex: Int) -> Int {
    guard !headings.isEmpty else { return 0 }
    var indexById: [String: Int] = [:]
    indexById.reserveCapacity(posts.count)
    for (index, post) in posts.enumerated() {
        indexById[post.id] = index
    }
    let limit = visibleIndex
    return headings.indices.last(where: { i in
        (indexById[headings[i].id] ?? Int.max) <= limit
    }) ?? 0
}
